qiskit/primitives/sampler.py

```diff
@@ -0,0 +1,107 @@
+
+
+class Sampler(BaseSampler):
+    """
+    Sampler class
+    """
+
+    def __init__(
+        self,
+        circuits: QuantumCircuit | Iterable[QuantumCircuit],
+        parameters: Iterable[Iterable[Parameter]] | None = None,
+    ):
+        """
+        Args:
+            circuits: circuits to be executed
+
+        Raises:
+            QiskitError: if some classical bits are not used for measurements.
+        """
+        if isinstance(circuits, QuantumCircuit):
+            circuits = [circuits]
+        circuits = [init_circuit(circuit) for circuit in circuits]
+        q_c_mappings = [final_measurement_mapping(circuit) for circuit in circuits]
+        self._qargs_list = []
+        for circuit, q_c_mapping in zip(circuits, q_c_mappings):
+            if set(range(circuit.num_clbits)) != set(q_c_mapping.values()):
+                raise QiskitError(
+                    "some classical bits are not used for measurements."
+                    f" the number of classical bits {circuit.num_clbits},"
+                    f" the used classical bits {set(q_c_mapping.values())}."
+                )
+            c_q_mapping = sorted((c, q) for q, c in q_c_mapping.items())
+            self._qargs_list.append([q for _, q in c_q_mapping])
+        circuits = [circuit.remove_final_measurements(inplace=False) for circuit in circuits]
+        super().__init__(circuits, parameters)
+        self._is_closed = False
+
+    def __call__(
+        self,
+        circuits: Sequence[int] | None = None,
+        parameters: Sequence[Sequence[float]] | None = None,
+        **run_options,
+    ) -> SamplerResult:
+        if self._is_closed:
+            raise QiskitError("The primitive has been closed.")
+
+        if circuits is None and parameters is not None and len(self._circuits) == 1:
+            circuits = [0] * len(parameters)
+        if circuits is None:
+            circuits = list(range(len(self._circuits)))
+        if parameters is None:
+            parameters = [[] for _ in self._circuits]
+        if len(circuits) != len(parameters):
+            raise QiskitError(
+                f"The number of circuits ({len(circuits)}) does not match "
+                f"the number of parameter sets ({len(parameters)})."
+            )
+
+        bound_circuits_qargs = []
+        for i, value in zip(circuits, parameters):
+            if len(value) != len(self._parameters[i]):
+                raise QiskitError(
+                    f"The number of values ({len(value)}) does not match "
+                    f"the number of parameters ({len(self._parameters[i])})."
+                )
+            bound_circuits_qargs.append(
+                (
+                    self._circuits[i].bind_parameters(dict(zip(self._parameters[i], value))),
+                    self._qargs_list[i],
+                )
+            )
+        probabilities = [
+            Statevector(circ).probabilities(qargs=qargs) for circ, qargs in bound_circuits_qargs
```

Comment:
I'm not sure that that should be a part of `quantum_info` itself (because it involves knowledge of a circuit), but it feels absolutely fine for `BasicAer` to be rewritten to use `quantum_info` underneath, and then for `BasicAer` to handle the mid-circuit measurement.

That way, BasicAer can handle partitioning the circuit around measurement components (since it's above circuits in the stack), and then can use `quantum_info` as worker routines to evaluate the quantum atoms.  The general task of handling classical operations (not just measurement) is going to be far more complex in the future, and it's definitely in the realm of a provider/backend.